pennylane/qchem/vibrational/vscf.py

```diff
@@ -0,0 +1,487 @@
+                    im = excitation_indices[m_idx]
+                    jm = excitation_indices[m_idx + len(modal_indices)]
+                    if im >= modals[m] or jm >= modals[m]:
+                        exc_in_modals = False
+                        break
+                if exc_in_modals:
+                    active_ham_terms[active_num] = [h_val, modal_indices, excitation_indices]
+                    active_num += 1
+
+    active_mode_terms = {
+        m: [t for t in range(active_num) if m in active_ham_terms[t][1]] for m in range(nmodes)
+    }
+
+    return active_ham_terms, active_mode_terms, active_num
+
+
+def _fock_energy(h_mat, active_ham_terms, active_mode_terms, modals, mode_rots):
+    r"""Calculates vibrational energy.
+
+    Args:
+        h_mat (Array[Array[float]]): the Hamiltonian matrix
+        active_ham_terms (dict): A dictionary containing the active Hamiltonian terms
+        active_mode_terms (dict): A list containing the active Hamiltonian terms for all vibrational modes.
+        modals (list[int]): A list containing the maximum number of modals to consider for each mode.
+        mode_rots (List[TensorLike[float]]): A list of rotation matrices for each vibrational mode.
+
+    Returns:
+        float: vibrational energy
+
+    """
+    nmodes = h_mat.shape[1]
+    e0s = np.zeros(nmodes)
+    for mode in range(nmodes):
+        fock_mat = _build_Fock(
+            mode, active_ham_terms, active_mode_terms[mode], modals, h_mat, mode_rots[mode]
+        )
+        e0s[mode] = fock_mat[0, 0]
+
+    return np.sum(e0s)
+
+
+def _vscf(h_integrals, modals, cutoff, tol=1e-8, max_iters=10000):
+    r"""Performs the VSCF calculation.
+
+    Args:
+        h_integrals (list(TensorLike[float])): a list containing Hamiltonian integral matrices
+        modals (list(int)): A list containing the maximum number of modals to consider for each mode.
+        cutoff (float): threshold value for including matrix elements into operator.
+        tol (float): convergence tolerance for vscf calculation
+        max_iters (int): Maximum number of iterations for vscf to converge.
+
+    Returns:
+        tuple : A tuple of the following:
+         - float: vscf energy
+         - List[TensorLike[float]]: A list of rotation matrices for all vibrational modes.
+
+    """
+
+    nmodes = np.shape(h_integrals[0])[0]
+
+    active_ham_terms, active_mode_terms, active_num = _find_active_terms(
+        h_integrals, modals, cutoff
+    )
+    mode_rots = [np.eye(modals[mode]) for mode in range(nmodes)]
+    h_mat = np.zeros((active_num, nmodes))
+    for mode in range(nmodes):
+        h_mat = _update_h(
+            h_mat, mode, active_ham_terms, mode_rots[mode], active_mode_terms[mode], modals
+        )
+
+    e0 = _fock_energy(h_mat, active_ham_terms, active_mode_terms, modals, mode_rots)
+
+    enew = e0 + 2 * tol
+    curr_iter = 0
+    while curr_iter <= max_iters and np.abs(enew - e0) > tol:
```

Comment:
Prefer for loop and break statement instead. I think in python for loops are more performant in general.